subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfigurationContainer.java

```diff
@@ -352,6 +375,122 @@ private <T extends Configuration> NamedDomainObjectProvider<T> registerConfigura
         return configuration;
     }
 
+    private static void emitConfigurationExistsDeprecation(String name) {
+        DeprecationLogger.deprecateBehaviour("The configuration " + name + " was created explicitly. This configuration name is reserved for creation by Gradle.")
+            .withAdvice(String.format("Do not create a configuration with the name %s.", name))
+            .willBeRemovedInGradle9()
+            .withUpgradeGuideSection(8, "configurations_allowed_usage")
+            .nagUser();
+    }
+
+    private static void validateNameIsAllowed(String name) {
+        if (RESERVED_NAMES_FOR_DETACHED_CONFS.matcher(name).matches()) {
+            DeprecationLogger.deprecateAction("Creating a configuration with a name that starts with 'detachedConfiguration'")
+                    .withAdvice(String.format("Use a different name for the configuration '%s'.", name))
+                    .willBeRemovedInGradle9()
+                    .withUpgradeGuideSection(8, "reserved_configuration_names")
+                    .nagUser();
+        }
+    }
+
+    /**
+     * Validates and ensures that the allowed usage of an existing configuration is consistent with the given expected usage.
+     *
+     * This method will emit a detailed deprecation method with suggestions if the usage is inconsistent.  It will then attempt to mutate
+     * the usage to match the expectation.  If the usage cannot be mutated, it will throw an exception.
+     *
+     * Does <strong>NOT</strong> check anything to do with deprecated usage.
+     *
+     * @param confName the name of the configuration
+     * @param expectedUsage the expectedUsage defining the usage the configuration should allow
+     */
+    private void validateExistingUsageIsConsistent(String confName, ConfigurationRole expectedUsage) {
+        DefaultConfiguration conf = (DefaultConfiguration) getByName(confName);
+        if (!expectedUsage.isUsageConsistentWithRole(conf)) {
```

Comment:
I've honestly never thought about avoiding leading indentation as improving readability.  This adds length, I think multiple exit points from a method are also not good, and I like keeping the nested blocks as context - now the reader has to remember they've already checked if usage is consistent when reading the bottom of the method (which probably means the method itself is too long).

However, there is a good amount of horizontal scrolling here with the messages.  I'll make the change in this case, but I don't really see this as a general principle - it would have to be a case by case thing, and (if this class wasn't already such a beast) extracting helper methods would probably be better.